language/vm/vm-runtime/src/move_vm.rs

```diff
@@ -72,17 +90,148 @@ impl VMExecutor for MoveVM {
+                    let outs = vm.rent(|runtime| {
+                        execute_user_transaction_block(txns, runtime, &mut data_cache, state_view)
+                    });
+                    result.append(&mut outs?);
+                }
+                TransactionBlock::BlockPrologue(block_metadata) => {
+                    result.push(vm.rent(|runtime| {
+                        process_block_metadata(block_metadata, runtime, state_view, &mut data_cache)
+                    }))
+                }
+                // TODO: Implement the logic for processing writeset transactions.
+                TransactionBlock::WriteSet(_) => unimplemented!(""),
+            }
+        }
+
+        Ok(result)
+    }
+}
+
+/// Transactions divided by transaction flow.
+/// Transaction flows are different across different types of transactions.
+pub enum TransactionBlock {
+    UserTransaction(Vec<SignedTransaction>),
+    WriteSet(WriteSet),
+    BlockPrologue(BlockMetadata),
+}
+
+pub fn chunk_block_transactions(txns: Vec<Transaction>) -> Vec<TransactionBlock> {
+    let mut blocks = vec![];
+    let mut buf = vec![];
+    for txn in txns {
+        match txn {
+            Transaction::BlockMetadata(data) => {
+                if !buf.is_empty() {
+                    blocks.push(TransactionBlock::UserTransaction(buf));
+                    buf = vec![];
+                }
+                blocks.push(TransactionBlock::BlockPrologue(data));
+            }
+            Transaction::WriteSet(ws) => {
+                if !buf.is_empty() {
+                    blocks.push(TransactionBlock::UserTransaction(buf));
+                    buf = vec![];
+                }
+                blocks.push(TransactionBlock::WriteSet(ws));
+            }
+            Transaction::UserTransaction(txn) => {
+                buf.push(txn);
+            }
+        }
+    }
+    if !buf.is_empty() {
+        blocks.push(TransactionBlock::UserTransaction(buf));
     }
+    blocks
+}
+
+/// Determine if a transaction is valid. Will return `None` if the transaction is accepted,
+/// `Some(Err)` if the VM rejects it, with `Err` as an error code. We verify the following
+/// items:
+/// 1. The signature on the `SignedTransaction` matches the public key included in the
+///    transaction
+/// 2. The script to be executed is in the whitelist.
+/// 3. Invokes `LibraAccount.prologue`, which checks properties such as the transaction has the
+/// right sequence number and the sender has enough balance to pay for the gas. 4.
+/// Transaction arguments matches the main function's type signature. 5. Script and modules
+/// in the transaction pass the bytecode static verifier.
+///
+/// Note: In the future. we may defer these checks to a later pass, as all the scripts we will
+/// execute are pre-verified scripts. And bytecode verification is expensive. Thus whether we
+/// want to perform this check here remains unknown.
+pub fn verify_transaction(
```

Comment:
this is the validation story so I moved it here because it's the entry point.
Could live elsewhere and it probably will